### PR TITLE
Case 22060 fix model

### PR DESCRIPTION
**Pull request by raveenajain**
https://highfidelity.manuscript.com/f/cases/22060/brainstem-gltf-model-broken#BugEvent.152435
Test PR: https://github.com/highfidelity/hifi_tests/pull/396

### CODE REVIEW

**Comment by raveenajain** [comment]
Test Plan: 

1. run Interface -> enter domain with minimal external influences
2. Edit -> Running Scripts 
3. Paste URL: https://raw.githubusercontent.com/highfidelity/hifi_tests/master/tests/content/entity/model/modelReaders/gltfReader/gltfTestSuite/brainstem/testAuto.js and/or https://raw.githubusercontent.com/highfidelity/hifi_tests/master/tests/content/entity/model/modelReaders/gltfReader/gltfTestSuite/embedded/brainstem/testAuto.js and/or https://raw.githubusercontent.com/highfidelity/hifi_tests/master/tests/content/entity/model/modelReaders/gltfReader/gltfTestSuite/embedded/brainstem/testAuto.js
4. Output should match expected output: https://github.com/highfidelity/hifi_tests/blob/master/tests/content/entity/model/modelReaders/gltfReader/gltfTestSuite/brainstem/ExpectedImage_00000.png

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18606/HighFidelity-Beta-PR15332-d1d999d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18606/HighFidelity-Beta-Interface-PR15332-d1d999d.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18606/HighFidelity-Beta-PR15332-d1d999d.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18606/HighFidelity-Quest-Beta-PR15332-d1d999d.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18606/HighFidelity-Beta-PR15332-d1d999d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18606/HighFidelity-Beta-Interface-PR15332-d1d999d.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18641/HighFidelity-Beta-PR15332-40fe156.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18641/HighFidelity-Beta-Interface-PR15332-40fe156.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18641/HighFidelity-Beta-PR15332-40fe156.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18641/HighFidelity-Beta-Interface-PR15332-40fe156.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18641/HighFidelity-Beta-PR15332-40fe156.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18641/HighFidelity-Quest-Beta-PR15332-40fe156.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18658/HighFidelity-Beta-PR15332-c970a58.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18658/HighFidelity-Beta-Interface-PR15332-c970a58.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18658/HighFidelity-Beta-PR15332-c970a58.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18658/HighFidelity-Beta-Interface-PR15332-c970a58.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18658/HighFidelity-Beta-PR15332-c970a58.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18658/HighFidelity-Quest-Beta-PR15332-c970a58.apk)

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18729/HighFidelity-Beta-PR15332-f458427.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18729/HighFidelity-Quest-Beta-PR15332-f458427.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18759/HighFidelity-Beta-PR15332-104ed75.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18759/HighFidelity-Beta-Interface-PR15332-104ed75.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18759/HighFidelity-Beta-PR15332-104ed75.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18759/HighFidelity-Quest-Beta-PR15332-104ed75.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18759/HighFidelity-Beta-PR15332-104ed75.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18759/HighFidelity-Beta-Interface-PR15332-104ed75.exe

File: libraries/fbx/src/GLTFSerializer.cpp [hunk 1]
```diff
@@ -753,107 +753,82 @@ void GLTFSerializer::getSkinInverseBindMatrices(std::vector<std::vector<float>>&
...
+            while (j < numNodes) {
+                int nextNode = sortedNodes[j];
+                parentIndex = parents[nextNode];
+                if (parentIndex == -1 || hasBeenSorted[parentIndex]) {
+                    // swap with currentNode
+                    hasBeenSorted[nextNode] = true;
+                    sortedNodes[i] = nextNode;
+                    sortedNodes[j] = currentNode;
+                    i++;
+                    break;
```
Comment (AndrewMeadows) [inline libraries/fbx/src/GLTFSerializer.cpp]: Remove this `break;` and replace it with `currentNode = sortedNodes[i];` which needs to be done after changing `i`.

File: libraries/fbx/src/GLTFSerializer.cpp [hunk 2]
```diff
@@ -952,10 +929,7 @@ bool GLTFSerializer::buildGeometry(HFMModel& hfmModel, const hifi::URL& url) {
                     }
                 }
                 HFMCluster root; 
-                root.jointIndex = nodeQueue[0];
-                if (root.jointIndex == -1) { 
-                    root.jointIndex = 0; 
-                }
+                root.jointIndex = sortedNodes[0];
```
Comment (AndrewMeadows) [inline libraries/fbx/src/GLTFSerializer.cpp]: This line seems wrong to me.  Why is it `root.jointIndex = sortedNodes[0]` and not `root.jointIndex = originalToNewNodeIndexMap[sortedNodes[0]];` ?

File: libraries/fbx/src/GLTFSerializer.cpp [hunk 3]
```diff
@@ -753,107 +753,82 @@ void GLTFSerializer::getSkinInverseBindMatrices(std::vector<std::vector<float>>&
...
     HFMJoint joint;
     joint.distanceToParent = 0;
     hfmModel.jointIndices["x"] = numNodes;
-    hfmModel.hasSkeletonJoints = false;
 
-    for (int nodeIndex : nodeQueue) {
+    for (int nodeIndex : sortedNodes) {
         auto& node = _file.nodes[nodeIndex];
 
         joint.parentIndex = -1;
```
Comment (AndrewMeadows) [inline libraries/fbx/src/GLTFSerializer.cpp]: The `if` check below is O(N).  I believe we can answer the question faster than that.  Hint: ask the `parents` array.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18769/HighFidelity-Beta-PR15332-2fab5a0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18769/HighFidelity-Beta-Interface-PR15332-2fab5a0.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18769/HighFidelity-Beta-PR15332-2fab5a0.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18769/HighFidelity-Quest-Beta-PR15332-2fab5a0.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18769/HighFidelity-Beta-PR15332-2fab5a0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18769/HighFidelity-Beta-Interface-PR15332-2fab5a0.exe

**Pull request merged**